von/strparse.py

```diff
@@ -1,9 +1,80 @@
+#!/usr/bin/env python3
+
 import re
 
+import pyperclip
+
+oper_macros = {
+    "\\floor": (" \\left\\lfloor ", " \\right\\rfloor "),
+    "\\ceil": (" \\left\\lceil ", " \\right\\rceil "),
+    "\\abs": (" \\left\\lvert ", " \\right\\lvert "),
+    "\\norm": (" \\left\\lVert ", " \\right\\lVert "),
+    "\\anbr": (" \\left\\langle ", " \\right\\rangle "),
+    "\\sbr": (" \\left[ ", " \\right] "),
+}
+derv_macros = {
+    "\\dd": "\\mathrm{d}",
+    "\\pd": "\\partial",
+}
+
+
+def bracket_cpos(text: str, brac: tuple, pos: int) -> int:
+    close_counter = 0
+    open_counter = 1
+    charpos = pos + 1
+    for i in text[pos + 1 :]:
+        if i == brac[0]:
+            open_counter += 1
+        elif i == brac[1]:
+            close_counter += 1
+        if close_counter == open_counter:
+            break
+        charpos += 1
+    return charpos
+
+
+def derv_demacro(text: str) -> str:
+    for key, value in derv_macros.items():
+        while text.find(key) != -1:
+            pos = text.find(key)
+            if text[pos + len(key)] == "{":
+                first_cur_cpos = bracket_cpos(text, ("{", "}"), pos + len(key))
+                second_cur_cpos = bracket_cpos(text, ("{", "}"), first_cur_cpos + 1)
+                inner_text_1 = text[pos + len(key) + 1 : first_cur_cpos]
+                inner_text_2 = text[first_cur_cpos + 2 : second_cur_cpos]
+                text = (
+                    text[:pos]
+                    + f"\\frac{{{value} {inner_text_1}}}{{{value} {inner_text_2}}}"
+                    + text[second_cur_cpos + 1 :]
+                )
+            elif text[pos + len(key)] == "[":
+                sq_cpos = bracket_cpos(text, ("[", "]"), pos + len(key))
+                first_cur_cpos = bracket_cpos(text, ("{", "}"), sq_cpos + 1)
+                second_cur_cpos = bracket_cpos(text, ("{", "}"), first_cur_cpos + 1)
+                func_deg = text[pos + len(key) + 1 : sq_cpos]
+                inner_text_1 = text[sq_cpos + 2 : first_cur_cpos]
+                inner_text_2 = text[first_cur_cpos + 2 : second_cur_cpos]
+                text = (
+                    text[:pos]
+                    + f"\\frac{{{value}^{{{func_deg}}} {inner_text_1}}}{{{value} {inner_text_2}^{{{func_deg}}}}}"
+                    + text[second_cur_cpos + 1 :]
+                )
+    return text
 
-# Demacro
-def demacro(text: str) -> str:
-    # TODO this doesn't quite work, but oh well
+
+def oper_demacro(text: str) -> str:
+    for key, value in oper_macros.items():
+        while text.find(key) != -1:
+            pos = text.find(key)
+            cur_cpos = bracket_cpos(text, ("{", "}"), pos + len(key))
+            inner_text_1 = text[pos + len(key) + 1 : cur_cpos]
+            text = (
+                text[:pos] + value[0] + inner_text_1 + value[1] + text[cur_cpos + 1 :]
+            )
+    return text
+
+
+def comm_demacro(text: str) -> str:
```

Comment:
i think this one should just be called demacro because it's used by other components